File: lib/node/vssh-app-fs/http.js
'use strict';

// Adaptador node:http do contrato de wire descrito em README.md.
//
// Nada aqui sabe qual app está montando o handler: recebe um objeto de ops (ops.js) e devolve uma
// função `(req, res) => Promise<boolean>` que diz se atendeu a requisição. Quem compõe as rotas é
// o server do app.

const crypto = require('node:crypto');

const { FsError, ENOENT, EACCES, EINVAL } = require('./paths');

const DEFAULT_MAX_BODY_BYTES = 64 * 1024 * 1024;

const CONTENT_TYPES = {
  png: 'image/png',
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  gif: 'image/gif',
  webp: 'image/webp',
  avif: 'image/avif',
  svg: 'image/svg+xml',
  bmp: 'image/bmp',
  ico: 'image/x-icon',
  pdf: 'application/pdf',
  mp3: 'audio/mpeg',
  ogg: 'audio/ogg',
  wav: 'audio/wav',
  m4a: 'audio/mp4',
  mp4: 'video/mp4',
  webm: 'video/webm',
  mov: 'video/quicktime',
  txt: 'text/plain; charset=utf-8',
  md: 'text/markdown; charset=utf-8',
  json: 'application/json; charset=utf-8',
  csv: 'text/csv; charset=utf-8',
};

function contentTypeFor(p) {
  const dot = p.lastIndexOf('.');
  const ext = dot === -1 ? '' : p.slice(dot + 1).toLowerCase();
  return CONTENT_TYPES[ext] || 'application/octet-stream';
}

// Errno do Node traduzido em status. O que fica de fora cai em 500 de propósito — 500 significa
// "o servidor não sabe o que aconteceu", e essa distinção só é útil se a lista do que ele SABE for
// honesta. Antes daqui só havia três entradas, e um EISDIR virava 500: um pedido inválido do
// cliente aparecia como defeito do backend, que é o diagnóstico errado.
const STATUS_BY_CODE = {
  [ENOENT]: 404,
  [EACCES]: 403,
  EPERM: 403,
  [EINVAL]: 400,
  EISDIR: 400,
  ENOTDIR: 400,
  ENAMETOOLONG: 400,
  ELOOP: 400,
  EEXIST: 409,
  // ENOSPC, EIO, EMFILE, EROFS e erro de programação continuam em 500: são do servidor mesmo.
};

function statusFor(err) {
  return STATUS_BY_CODE[err && err.code] || 500;
}

function sendJson(res, status, payload) {
  const body = Buffer.from(JSON.stringify(payload), 'utf8');
  res.writeHead(status, {
    'Content-Type': 'application/json; charset=utf-8',
    'Content-Length': body.length,
    'Cache-Control': 'no-store',
  });
  res.end(body);
}

function sendError(res, err, op) {
  const status = statusFor(err);
  sendJson(res, status, {
    ok: false,
    error: {
      code: err.code || 'EUNKNOWN',
      message: err.message || String(err),
      // Sem isto o cliente só sabe que "uma op falhou" — e o console do navegador mostra a URL do
      // mount, que é a mesma para todas.
      ...(op ? { op } : {}),
    },
  });
}

// Só o que cai em 500 leva `stack`. Nos erros classificados a pilha não acrescenta nada — é sempre
// a mesma —, e nos 500 é a única coisa que diz onde o backend tropeçou.
//
// `expected` marca ENOENT: "não existe" é resposta de rotina para quem sonda antes de criar, e sem
// essa marca a linha diz `op-failed` e parece defeito. Foi o que fez três sondagens normais
// passarem por sintoma durante a depuração.
function failureEvent(event, err, extra) {
  const classified = Boolean(STATUS_BY_CODE[err && err.code]);
  return {
    event,
    ...extra,
    code: err.code || 'EUNKNOWN',
    message: err.message,
    ...(err && err.code === ENOENT ? { expected: true } : {}),
    ...(classified ? {} : { stack: err.stack }),
  };
}

function readBody(req, limit) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    let total = 0;
    req.on('data', (chunk) => {
      total += chunk.length;
      if (total > limit) {
        reject(Object.assign(new Error('corpo grande demais'), { code: EINVAL }));
        req.destroy();
        return;
      }
      chunks.push(chunk);
    });
    req.on('end', () => resolve(Buffer.concat(chunks)));
    req.on('error', reject);
  });
}

// Comparação de tamanho fixo: hash dos dois lados antes de comparar, para não vazar o prefixo
// coincidente pelo tempo nem tropeçar em comprimentos diferentes.
function tokenMatches(expected, received) {
  if (typeof received !== 'string' || received.length === 0) return false;
  const a = crypto.createHash('sha256').update(expected).digest();
  const b = crypto.createHash('sha256').update(received).digest();
  return crypto.timingSafeEqual(a, b);
}

/**
 * @param {object} options
 * @param {object} options.fs                  objeto devolvido por createAppFs
 * @param {string} [options.mountPath]         rota do RPC JSON (default '/api/fs')
 * @param {string} [options.assetsPrefix]      rota dos binários (default '/assets/')
 * @param {string|null} [options.requireToken] se string, exige X-Vssh-App-Token igual
 * @param {number} [options.maxBodyBytes]
 * @param {(event: object) => void} [options.onWarn]
 */
function createFsHandler(options) {
  const graphFs = options.fs;
  const mountPath = options.mountPath || '/api/fs';
  const assetsPrefix = options.assetsPrefix || '/assets/';
  const requireToken = options.requireToken || null;
  const maxBodyBytes = options.maxBodyBytes ?? DEFAULT_MAX_BODY_BYTES;
  const onWarn = options.onWarn || (() => {});

  const OPS = {
    'exists': (p) => graphFs.exists(p.path),
    'stat': (p) => graphFs.stat(p.path),
    'read-file': (p) => graphFs.readFile(p.path),
    'write-file': (p) => graphFs.writeFile(p.path, p.content ?? ''),
    'mkdir': (p) => graphFs.mkdir(p.path ?? p.dir),
    'mkdir-recur': (p) => graphFs.mkdirRecur(p.path ?? p.dir),
    'readdir': (p) => graphFs.readdir(p.path ?? p.dir, { applyIgnore: !!p.applyIgnore }),
    'unlink': (p) => graphFs.unlink(p.path, { recycle: p.recycle !== false }),
    'rename': (p) => graphFs.rename(p.from ?? p.old, p.to ?? p.new),
    'copy': (p) => graphFs.copy(p.from ?? p.old, p.to ?? p.new),
    'open-dir': (p) => graphFs.openDir(p.path ?? p.dir),
    'get-files': (p) => graphFs.getFiles(p.path ?? p.dir),
  };

  async function handleRpc(req, res) {
    if (req.method !== 'POST') {
      sendJson(res, 405, { ok: false, error: { code: 'EMETHOD', message: 'use POST' } });
      return;
    }
    let payload;
    try {
      const raw = await readBody(req, maxBodyBytes);
      payload = JSON.parse(raw.toString('utf8'));
    } catch (err) {
      sendError(res, Object.assign(err, { code: err.code || EINVAL }));
      return;
    }
    const op = OPS[payload && payload.op];
    if (!op) {
      sendJson(res, 400, {
        ok: false,
        error: { code: EINVAL, message: `op desconhecida: ${payload && payload.op}` },
      });
      return;
    }
    try {
      sendJson(res, 200, { ok: true, result: await op(payload) });
    } catch (err) {
      onWarn(failureEvent('op-failed', err, { op: payload.op, path: payload.path }));
      sendError(res, err, payload.op);
    }
  }

  async function handleWriteBinary(req, res, url) {
    if (req.method !== 'POST' && req.method !== 'PUT') {
      sendJson(res, 405, { ok: false, error: { code: 'EMETHOD', message: 'use POST' } });
      return;
    }
    const target = url.searchParams.get('path');
    try {
      const body = await readBody(req, maxBodyBytes);
      sendJson(res, 200, { ok: true, result: await graphFs.writeFile(target, body) });
    } catch (err) {
      onWarn(failureEvent('write-binary-failed', err, { path: target }));
      sendError(res, err, 'write-binary');
    }
  }

  async function handleAsset(req, res, relative) {
    if (req.method !== 'GET' && req.method !== 'HEAD') {
      sendJson(res, 405, { ok: false, error: { code: 'EMETHOD', message: 'use GET' } });
      return;
    }
    // `%` malformado faz decodeURIComponent lançar URIError, que não tem `.code` — sem este
    // tratamento ele caía no 500 genérico, quando é URL inválida do cliente, ou seja 400.
    let decoded;
    try {
      decoded = decodeURIComponent(relative);
    } catch {
      sendError(res, new FsError(EINVAL, `caminho de asset inválido: ${relative}`));
      return;
    }

    let file;
    try {
      file = await graphFs.openRead(decoded);
    } catch (err) {
      sendError(res, err);
      return;
    }

    const lastModified = new Date(file.mtime).toUTCString();
    const headers = {
      'Content-Type': contentTypeFor(file.path),
      'Last-Modified': lastModified,
      'Cache-Control': 'no-cache',
      'Accept-Ranges': 'bytes',
    };

    // Comparação exata de string, não parse de data: o valor veio deste mesmo servidor. Faltava
    // aqui e existia no static-spa — assimetria sem motivo entre as duas peças.
    if (req.headers['if-modified-since'] === lastModified && !req.headers.range) {
      res.writeHead(304, { 'Last-Modified': lastModified, 'Cache-Control': 'no-cache' });
      res.end();
      return;
    }

    // Range de um intervalo só — o pdf.js do Logseq usa isso para não baixar o PDF inteiro.
    const range = /^bytes=(\d*)-(\d*)$/.exec(req.headers.range || '');
    if (range && file.size > 0) {
      const [, rawStart, rawEnd] = range;
      let start = rawStart === '' ? file.size - Number(rawEnd) : Number(rawStart);
      let end = rawStart === '' || rawEnd === '' ? file.size - 1 : Number(rawEnd);
      start = Math.max(0, start);
      end = Math.min(file.size - 1, end);
      if (Number.isNaN(start) || Number.isNaN(end) || start > end) {
        res.writeHead(416, { 'Content-Range': `bytes */${file.size}` });
        res.end();
        return;
      }
      headers['Content-Range'] = `bytes ${start}-${end}/${file.size}`;
      headers['Content-Length'] = end - start + 1;
      res.writeHead(206, headers);
      if (req.method === 'HEAD') return res.end();
      return void file.stream({ start, end }).pipe(res);
    }

    headers['Content-Length'] = file.size;
    res.writeHead(200, headers);
    if (req.method === 'HEAD') return res.end();
    file.stream().pipe(res);
  }

  return async function handle(req, res, url) {
    const pathname = url.pathname;
    const isRpc = pathname === mountPath;
    const isWriteBinary = pathname === mountPath + '/write-binary';
    const isAsset = pathname.startsWith(assetsPrefix);
    if (!isRpc && !isWriteBinary && !isAsset) return false;

    if (requireToken && !tokenMatches(requireToken, req.headers['x-vssh-app-token'])) {
      // Requisição que não veio pelo proxy autenticado do portal. A porta é loopback, mas outro
      // processo do mesmo usuário Linux alcança — e este handler dá acesso a arquivos.
      onWarn({ event: 'token-rejected', pathname });
      sendJson(res, 403, { ok: false, error: { code: EACCES, message: 'token ausente ou inválido' } });
      return true;
    }

    if (isRpc) await handleRpc(req, res);
    else if (isWriteBinary) await handleWriteBinary(req, res, url);
    else await handleAsset(req, res, pathname.slice(assetsPrefix.length));
    return true;
  };
}

module.exports = { createFsHandler, contentTypeFor, tokenMatches, DEFAULT_MAX_BODY_BYTES };

File: lib/python/vssh_app_toolkit/fs/http.py
"""O adaptador do contrato de wire para o `http.server` da stdlib.

O par do `lib/node/vssh-app-fs/http.js`. O contrato é o MESMO — um RPC JSON num POST, mais uma rota
de binário com `Range` — porque o frontend é o mesmo em qualquer runtime. Um app que trocasse o
backend de Node para Python não deveria precisar tocar numa linha do próprio JavaScript.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import re
from email.utils import formatdate
from urllib.parse import unquote, urlsplit, parse_qs

from .paths import EACCES, EINVAL, ENOENT, ErroFs

__all__ = ["criar_handler_fs", "tipo_de_conteudo", "token_confere", "MAX_BYTES_DE_CORPO_PADRAO"]

MAX_BYTES_DE_CORPO_PADRAO = 64 * 1024 * 1024

# Mapa de conteúdo de DADOS do app (o que o usuário guarda), e não de bundle web — este é o irmão
# do mapa do `spa.py`, e eles são separados de propósito: as duas peças são independentes.
_TIPOS = {
    "md": "text/markdown; charset=utf-8", "txt": "text/plain; charset=utf-8",
    "json": "application/json; charset=utf-8", "edn": "application/edn; charset=utf-8",
    "org": "text/plain; charset=utf-8", "csv": "text/csv; charset=utf-8",
    "html": "text/html; charset=utf-8", "css": "text/css; charset=utf-8",
    "js": "text/javascript; charset=utf-8",
    "png": "image/png", "jpg": "image/jpeg", "jpeg": "image/jpeg", "gif": "image/gif",
    "webp": "image/webp", "svg": "image/svg+xml", "ico": "image/x-icon", "bmp": "image/bmp",
    "pdf": "application/pdf",
    "mp3": "audio/mpeg", "wav": "audio/wav", "ogg": "audio/ogg", "m4a": "audio/mp4",
    "mp4": "video/mp4", "webm": "video/webm", "mov": "video/quicktime",
    "zip": "application/zip",
}

_STATUS = {ENOENT: 404, EACCES: 403, EINVAL: 400, "EEXIST": 409, "EMETHOD": 405}


def tipo_de_conteudo(caminho):
    _, _, ext = caminho.rpartition(".")
    return _TIPOS.get(ext.lower(), "application/octet-stream")


def token_confere(esperado, recebido):
    """Comparação resistente a timing.

    Hash dos dois lados antes de comparar: sem isso, o tempo de resposta vaza o PREFIXO acertado, e
    comprimentos diferentes vazam o tamanho. É exportada porque o gate da lib já fazia isto
    enquanto a cola de cada app comparava com `!=` — lib rigorosa não adianta se quem a monta não
    for.
    """
    if not isinstance(recebido, str) or recebido == "":
        return False
    a = hashlib.sha256(esperado.encode("utf-8")).digest()
    b = hashlib.sha256(recebido.encode("utf-8")).digest()
    return hmac.compare_digest(a, b)


def _mandar_json(handler, status, corpo):
    dados = json.dumps(corpo, ensure_ascii=False, default=str, separators=(",", ":")).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Content-Length", str(len(dados)))
    handler.end_headers()
    if handler.command != "HEAD":
        handler.wfile.write(dados)


def _mandar_erro(handler, err, op=None):
    codigo = getattr(err, "codigo", None) or "EINTERNAL"
    # 500 é reservado para o que é DEFEITO NOSSO. Um caminho que não existe, um pedido malformado
    # ou um acesso negado são respostas corretas a pedidos errados — e chamá-los de 500 faz o
    # monitoramento acordar alguém por um clique num arquivo apagado.
    status = _STATUS.get(codigo, 500)
    corpo = {"ok": False, "error": {"code": codigo, "message": str(err)}}
    if op:
        corpo["error"]["op"] = op
    _mandar_json(handler, status, corpo)


def _ler_corpo(handler, maximo):
    tamanho = int(handler.headers.get("Content-Length") or 0)
    if tamanho > maximo:
        raise ErroFs(EINVAL, f"corpo maior que o teto ({tamanho} > {maximo})")
    return handler.rfile.read(tamanho) if tamanho else b""


def criar_handler_fs(fs, mount_path="/api/fs", assets_prefix="/assets/", require_token=None,
                     max_body_bytes=None, ao_avisar=None):
    """Devolve `servir(handler) -> bool`. `True` quer dizer que a requisição foi atendida."""
    maximo = MAX_BYTES_DE_CORPO_PADRAO if max_body_bytes is None else max_body_bytes
    avisar = ao_avisar or (lambda _e: None)

    OPS = {
        "exists": lambda p: fs.exists(p.get("path")),
        "stat": lambda p: fs.stat(p.get("path")),
        "read-file": lambda p: fs.read_file(p.get("path")),
        "write-file": lambda p: fs.write_file(p.get("path"), p.get("content") or ""),
        "mkdir": lambda p: fs.mkdir(p.get("path") or p.get("dir")),
        "mkdir-recur": lambda p: fs.mkdir_recur(p.get("path") or p.get("dir")),
        "readdir": lambda p: fs.readdir(p.get("path") or p.get("dir"),
                                        aplicar_ignore=bool(p.get("applyIgnore"))),
        "unlink": lambda p: fs.unlink(p.get("path"), reciclar=p.get("recycle") is not False),
        "rename": lambda p: fs.rename(p.get("from") or p.get("old"), p.get("to") or p.get("new")),
        "copy": lambda p: fs.copy(p.get("from") or p.get("old"), p.get("to") or p.get("new")),
        "open-dir": lambda p: fs.open_dir(p.get("path") or p.get("dir")),
        "get-files": lambda p: fs.get_files(p.get("path") or p.get("dir")),
    }

    def _rpc(handler):
        if handler.command != "POST":
            _mandar_json(handler, 405, {"ok": False, "error": {"code": "EMETHOD", "message": "use POST"}})
            return
        try:
            carga = json.loads(_ler_corpo(handler, maximo).decode("utf-8") or "{}")
        except ErroFs as err:
            _mandar_erro(handler, err)
            return
        except (ValueError, UnicodeDecodeError) as err:
            _mandar_erro(handler, ErroFs(EINVAL, f"corpo não é JSON: {err}"))
            return

        op = OPS.get(carga.get("op")) if isinstance(carga, dict) else None
        if not op:
            nome = carga.get("op") if isinstance(carga, dict) else None
            _mandar_json(handler, 400, {"ok": False,
                                        "error": {"code": EINVAL, "message": f"op desconhecida: {nome}"}})
            return
        try:
            _mandar_json(handler, 200, {"ok": True, "result": op(carga)})
        except ErroFs as err:
            avisar({"event": "op-failed", "op": carga.get("op"), "path": carga.get("path"),
                    "code": err.codigo, "message": str(err)})
            _mandar_erro(handler, err, carga.get("op"))
        except OSError as err:
            avisar({"event": "op-failed", "op": carga.get("op"), "path": carga.get("path"),
                    "message": str(err)})
            _mandar_erro(handler, err, carga.get("op"))

    def _escrever_binario(handler, consulta):
        if handler.command not in ("POST", "PUT"):
            _mandar_json(handler, 405, {"ok": False, "error": {"code": "EMETHOD", "message": "use POST"}})
            return
        alvo = (parse_qs(consulta).get("path") or [None])[0]
        try:
            corpo = _ler_corpo(handler, maximo)
            _mandar_json(handler, 200, {"ok": True, "result": fs.write_file(alvo, corpo)})
        except (ErroFs, OSError) as err:
            avisar({"event": "write-binary-failed", "path": alvo, "message": str(err)})
            _mandar_erro(handler, err, "write-binary")

    def _asset(handler, relativo):
        if handler.command not in ("GET", "HEAD"):
            _mandar_json(handler, 405, {"ok": False, "error": {"code": "EMETHOD", "message": "use GET"}})
            return
        try:
            decodificado = unquote(relativo, errors="strict")
        except (UnicodeDecodeError, ValueError):
            _mandar_erro(handler, ErroFs(EINVAL, f"caminho de asset inválido: {relativo}"))
            return

        try:
            arquivo = fs.open_read(decodificado)
        except (ErroFs, OSError) as err:
            _mandar_erro(handler, err)
            return

        ultima = formatdate(arquivo["mtime"] / 1000.0, usegmt=True)
        faixa_pedida = handler.headers.get("Range")

        if handler.headers.get("If-Modified-Since") == ultima and not faixa_pedida:
            handler.send_response(304)
            handler.send_header("Last-Modified", ultima)
            handler.send_header("Cache-Control", "no-cache")
            handler.end_headers()
            return

        tamanho = arquivo["size"]
        inicio, fim = 0, tamanho - 1
        parcial = False

        # Range de UM intervalo só — é o que um leitor de PDF ou um `<video>` usa para não baixar o
        # arquivo inteiro. Multipart não vale o custo aqui, e nenhum cliente do ambiente o pede.
        m = re.fullmatch(r"bytes=(\d*)-(\d*)", faixa_pedida or "")
        if m and tamanho > 0:
            cru_inicio, cru_fim = m.group(1), m.group(2)
            try:
                if cru_inicio == "":
                    inicio = tamanho - int(cru_fim)
                    fim = tamanho - 1
                else:
                    inicio = int(cru_inicio)
                    fim = int(cru_fim) if cru_fim else tamanho - 1
            except ValueError:
                inicio, fim = 0, -1
            inicio = max(0, inicio)
            fim = min(tamanho - 1, fim)
            if inicio > fim:
                handler.send_response(416)
                handler.send_header("Content-Range", f"bytes */{tamanho}")
                handler.end_headers()
                return
            parcial = True

        comprimento = fim - inicio + 1 if tamanho else 0
        handler.send_response(206 if parcial else 200)
        handler.send_header("Content-Type", tipo_de_conteudo(arquivo["path"]))
        handler.send_header("Last-Modified", ultima)
        handler.send_header("Cache-Control", "no-cache")
        handler.send_header("Accept-Ranges", "bytes")
        if parcial:
            handler.send_header("Content-Range", f"bytes {inicio}-{fim}/{tamanho}")
        handler.send_header("Content-Length", str(comprimento))
        handler.end_headers()
        if handler.command == "HEAD":
            return
        with arquivo["abrir"]() as fh:
            fh.seek(inicio)
            restante = comprimento
            while restante > 0:
                pedaco = fh.read(min(64 * 1024, restante))
                if not pedaco:
                    break
                handler.wfile.write(pedaco)
                restante -= len(pedaco)

    def servir(handler):
        partes = urlsplit(handler.path)
        caminho = partes.path
        e_rpc = caminho == mount_path
        e_binario = caminho == mount_path + "/write-binary"
        e_asset = caminho.startswith(assets_prefix)
        if not (e_rpc or e_binario or e_asset):
            return False

        if require_token and not token_confere(require_token, handler.headers.get("X-Vssh-App-Token")):
            # Requisição que não veio pelo proxy autenticado do portal. O socket é 0600 do dono,
            # mas outro processo do MESMO usuário Linux o alcança — e este handler dá acesso a
            # arquivos.
            avisar({"event": "token-rejected", "pathname": caminho})
            _mandar_json(handler, 403, {"ok": False,
                                        "error": {"code": EACCES, "message": "token ausente ou inválido"}})
            return True

        if e_rpc:
            _rpc(handler)
        elif e_binario:
            _escrever_binario(handler, partes.query)
        else:
            _asset(handler, caminho[len(assets_prefix):])
        return True

    return servir
